lib/upipe-modules/upipe_dup.c

```diff
@@ -241,24 +238,18 @@ static void upipe_dup_input(struct upipe *upipe, struct uref *uref,
                             struct upump **upump_p)
 {
     struct upipe_dup *upipe_dup = upipe_dup_from_upipe(upipe);
-    struct uchain *uchain;
-    ulist_foreach (&upipe_dup->outputs, uchain) {
+
+    for (struct upipe *output = NULL;
+         ubase_check(upipe_dup_iterate_output(upipe, &output)) && output;) {
         struct upipe_dup_output *upipe_dup_output =
-            upipe_dup_output_from_uchain(uchain);
-        if (ulist_is_last(&upipe_dup->outputs, uchain)) {
-            upipe_dup_output_output(upipe_dup_output_to_upipe(upipe_dup_output),
-                                    uref, upump_p);
-            uref = NULL;
-        } else {
-            struct uref *new_uref = uref_dup(uref);
-            if (unlikely(new_uref == NULL)) {
-                uref_free(uref);
-                upipe_throw_fatal(upipe, UBASE_ERR_ALLOC);
-                return;
-            }
-            upipe_dup_output_output(upipe_dup_output_to_upipe(upipe_dup_output),
-                                    new_uref, upump_p);
+            upipe_dup_output_from_upipe(output);
+        struct uref *new_uref = uref_dup(uref);
```

Comment:
There is an extra uref_dup(), which is what the previous implementation intended to avoid. And it's called for every uref.